ethereumj-core/src/main/java/org/ethereum/listener/BlockReplay.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) [2016] [ <ether.camp> ]
+ * This file is part of the ethereumJ library.
+ *
+ * The ethereumJ library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ethereumJ library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the ethereumJ library. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.ethereum.listener;
+
+import org.apache.commons.collections4.queue.CircularFifoQueue;
+import org.ethereum.core.*;
+import org.ethereum.db.BlockStore;
+import org.ethereum.db.TransactionStore;
+import org.ethereum.net.eth.message.StatusMessage;
+import org.ethereum.net.message.Message;
+import org.ethereum.net.p2p.HelloMessage;
+import org.ethereum.net.rlpx.Node;
+import org.ethereum.net.server.Channel;
+import org.ethereum.util.FastByteComparisons;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Class capable of replaying stored blocks prior to 'going online' and
+ * notifying on newly imported blocks
+ *
+ * All other EthereumListener events are just forwarded to the supplied listener.
+ *
+ * Created by Anton Nashatyrev on 18.07.2016.
+ */
+public class BlockReplay extends EthereumListenerAdapter {
+    private static final Logger logger = LoggerFactory.getLogger("events");
+
+    BlockStore blockStore;
+    TransactionStore transactionStore;
+
+    EthereumListener listener;
+
+    long firstBlock;
+
+    boolean replayComplete = false;
+    Block lastReplayedBlock;
+    CircularFifoQueue<BlockSummary> onBlockBuffer = new CircularFifoQueue<>(192 * 2);
+
+    public BlockReplay(BlockStore blockStore, TransactionStore transactionStore, EthereumListener listener, long firstBlock) {
+        this.blockStore = blockStore;
+        this.transactionStore = transactionStore;
+        this.listener = listener;
+        this.firstBlock = firstBlock;
+    }
+
+    /**
+     * Replay blocks asynchronously
+     */
+    public void replayAsync() {
+        new Thread(this::replay).start();
+    }
+
+    /**
+     * Replay blocks synchronously
+     */
+    public void replay() {
+        long lastBlock = blockStore.getMaxNumber();
+        logger.info("Replaying blocks from " + firstBlock + ", current best block: " + lastBlock);
+        int cnt = 0;
+        long num = firstBlock;
+        while(true) {
```

Comment:
`while (!replayComplete)` is better if it does not break the logic (as I can see it does not)

Reply:
Sounds good